src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs

```diff
@@ -97,51 +98,65 @@ internal static SafeSslHandle AllocateSslContext(SslProtocols protocols, SafeX50
+                            }
                         }
                     }
-                }
 
-                context = SafeSslHandle.Create(innerContext, sslAuthenticationOptions.IsServer);
-                Debug.Assert(context != null, "Expected non-null return value from SafeSslHandle.Create");
-                if (context.IsInvalid)
-                {
-                    context.Dispose();
-                    throw CreateSslException(SR.net_allocate_ssl_context_failed);
-                }
+                    context = SafeSslHandle.Create(innerContext, sslAuthenticationOptions.IsServer);
+                    Debug.Assert(context != null, "Expected non-null return value from SafeSslHandle.Create");
+                    if (context.IsInvalid)
+                    {
+                        context.Dispose();
+                        throw CreateSslException(SR.net_allocate_ssl_context_failed);
+                    }
 
-                if (hasCertificateAndKey)
-                {
-                    bool hasCertReference = false;
-                    try
+                    if (hasCertificateAndKey)
                     {
-                        certHandle.DangerousAddRef(ref hasCertReference);
-                        using (X509Certificate2 cert = new X509Certificate2(certHandle.DangerousGetHandle()))
+                        bool hasCertReference = false;
+                        try
                         {
-                            using (X509Chain chain = TLSCertificateExtensions.BuildNewChain(cert, includeClientApplicationPolicy: false))
+                            certHandle.DangerousAddRef(ref hasCertReference);
+                            using (X509Certificate2 cert = new X509Certificate2(certHandle.DangerousGetHandle()))
                             {
-                                if (chain != null && !Ssl.AddExtraChainCertificates(context, chain))
-                                    throw CreateSslException(SR.net_ssl_use_cert_failed);
+                                using (X509Chain chain = TLSCertificateExtensions.BuildNewChain(cert, includeClientApplicationPolicy: false))
+                                {
+                                    if (chain != null && !Ssl.AddExtraChainCertificates(context, chain))
+                                        throw CreateSslException(SR.net_ssl_use_cert_failed);
+                                }
                             }
                         }
+                        finally
+                        {
+                            if (hasCertReference)
+                                certHandle.DangerousRelease();
+                        }
                     }
-                    finally
+
+                    context.AlpnHandle = alpnHandle;
+                }
+                catch
+                {
+                    if (alpnHandle.IsAllocated)
                     {
```

Comment:
Where will this get freed when there's no exception? This should clear after the handshake phase, I don't see the code in SecureChannel..

Reply:
What if you rehandshake? It's released when the ssl object is disposed

Reply:
Let me expand on this (was on the train). The GCHandle is not pinned so it really is just another object ref. The lifetime of this list is for the lifetime of the SslStream object anyway because its caught in the OptionsBag that is a member of SslState which is a member of SslStream. 

It won't cause the list to be held any longer than it currently is, and it won't take any extra memory, if a rehandshake takes place and the protocol match is recalled, it will be available otherwise you will have to make sure a new GCHandle is created at this time.

This is far superior to the current situation where the GCHandle is actually a pinned allocated buffer which will live for the SslStream lifetime. Because a pinned buffer will cause GC fragmentation which is not great.

Finally it is in the Dispose method of the Ssl object which is a SafeHandle so it will always be cleaned up.